crates/sui-transactional-test-runner/src/test_adapter.rs

```diff
@@ -740,6 +759,30 @@ impl<'a> MoveTestAdapter<'a> for SuiTestAdapter<'a> {
                     stop_line,
                     data,
                     |adapter, modules| {
+                        for (name, addr) in original_package_addrs {
+                            adapter
+                                .compiled_state()
+                                .named_address_mapping
+                                .insert(name.to_string(), addr)
+                                .unwrap_or_else(|| panic!("Internal error: expected dependency {name} in map when restoring address."));
+                        }
```

Comment:
So above we are updating addresses and doing a "push" before compilation, so we can compile successfully. This subsequent chunk is doing a "pop" before upgrading (restoring the updated addresses for upgrade/linking). Your question makes me think we might be able to avoid a push/pop on `named_address_mapping` but I didn't come up with a concrete solution for that.

We might be able to remove the need to reference or mutate `named_address_mapping` here, if we copy out and keep only the addresses needed for upgrading to a new map and then using that here instead of restoring `named_address_mapping`, if that's the direction you had in mind?